File: src/test/java/burrows/apps/example/gif/presentation/adapter/GifAdapterTest.java
package burrows.apps.example.gif.presentation.adapter;

import android.view.ViewGroup;
import android.widget.LinearLayout;
import burrows.apps.example.gif.data.rest.repository.ImageRepository;
import burrows.apps.example.gif.presentation.adapter.model.ImageInfo;
import com.bumptech.glide.GifRequestBuilder;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import test.RoboTestBase;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.atLeast;
import static org.mockito.Mockito.atLeastOnce;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.MockitoAnnotations.initMocks;

/**
 * @author <a href="mailto:jaredsburrows@gmail.com">Jared Burrows</a>
 */
@SuppressWarnings({"rawtypes", "unchecked"})
public final class GifAdapterTest extends RoboTestBase {
  private final ImageInfo imageInfo = new ImageInfo.Builder().url(STRING_UNIQUE).build();
  private final ImageInfo imageInfo2 = new ImageInfo.Builder().url(STRING_UNIQUE2).build();
  private final ImageInfo imageInfo3 = new ImageInfo.Builder().url(STRING_UNIQUE3).build();
  private GifAdapter.ViewHolder viewHolder;
  @Mock GifAdapter.OnItemClickListener onItemClickListener;
  @Mock GifRequestBuilder requestBuilder;
  ImageRepository spyImageDownloader;
  private GifAdapter sut;

  @Before @Override public void setUp() throws Exception {
    super.setUp();

    initMocks(this);

    spyImageDownloader = spy(new ImageRepository(context));
    sut = new GifAdapter(onItemClickListener, spyImageDownloader);
    sut.add(imageInfo);
    sut.add(imageInfo2);
    viewHolder = sut.onCreateViewHolder(new LinearLayout(context), 0);
  }

  @After @Override public void tearDown() throws Exception {
    super.tearDown();

    sut.clear();
  }

  @Test public void testOnCreateViewHolder() throws Exception {
    final ViewGroup parent = new ViewGroup(context) {
      @Override protected void onLayout(boolean b, int i, int i1, int i2, int i3) {
      }
    };

    assertThat(sut.onCreateViewHolder(parent, 0)).isInstanceOf(GifAdapter.ViewHolder.class);
  }

  @Test public void testOnBindViewHolderOnAdapterItemClick() {
    sut.clear();

    // must have one
    sut.add(imageInfo);
    sut.add(imageInfo2);
    sut.add(new ImageInfo());

    sut.onBindViewHolder(viewHolder, 0);

    assertThat(viewHolder.itemView.performClick()).isTrue();

    verify(spyImageDownloader, atLeastOnce()).load(any());
    verify(onItemClickListener).onClick(any());
  }

  @Test public void testGetItem() {
    sut.clear();

    final ImageInfo imageInfo = new ImageInfo();
    sut.add(imageInfo);

    assertThat(sut.getItem(0)).isEqualTo(imageInfo);
  }

  @Test public void onViewRecycled() throws Exception {
    sut.add(new ImageInfo());
    sut.onBindViewHolder(viewHolder, 0);

    sut.onViewRecycled(viewHolder);
  }

  @Test public void testGetItemCountShouldReturnCorrectValues() {
    assertThat(sut.getItemCount()).isEqualTo(2);
  }

  @Test public void testGetListCountShouldReturnCorrectValues() {
    assertThat(sut.getList()).isEqualTo(Arrays.asList(imageInfo, imageInfo2));
  }

  @Test public void testGetItemShouldReturnCorrectValues() {
    assertThat(sut.getItem(1)).isEqualTo(imageInfo2);
  }

  @Test public void testGetLocationShouldReturnCorrectValues() {
    assertThat(sut.getLocation(imageInfo2)).isEqualTo(1);
  }

  @Test public void testClearShouldClearAdapter() {
    sut.clear();

    assertThat(sut.getItemCount()).isEqualTo(0);
  }

  @Test public void testAddObjectShouldReturnCorrectValues() {
    sut.add(imageInfo3);

    assertThat(sut.getList()).isEqualTo(Arrays.asList(imageInfo, imageInfo2, imageInfo3));
  }

  @Test public void testAddCollectionShouldReturnCorrectValues() {
    final List<ImageInfo> imageInfos = Collections.singletonList(imageInfo3);

    sut.addAll(imageInfos);

    assertThat(sut.getList()).isEqualTo(Arrays.asList(imageInfo, imageInfo2, imageInfo3));
  }

  @Test public void testAddLocationObjectShouldReturnCorrectValues() {
    sut.add(0, new ImageInfo.Builder().url(STRING_UNIQUE3).build());

    assertThat(sut.getList()).isEqualTo(Arrays.asList(imageInfo3, imageInfo, imageInfo2));
  }

  @Test public void testRemoveLocationObjectShouldReturnCorrectValues() {
    sut.remove(0, imageInfo);

    assertThat(sut.getList()).isEqualTo(Collections.singletonList(imageInfo2));
  }

  @Test public void testRemoveObjectShouldReturnCorrectValues() {
    sut.remove(imageInfo);

    assertThat(sut.getList()).isEqualTo(Collections.singletonList(imageInfo2));
  }

  @Test public void testRemoveLocationShouldReturnCorrectValues() {
    sut.remove(0);

    assertThat(sut.getList()).isEqualTo(Collections.singletonList(imageInfo2));
  }
}

File: src/androidTest/java/burrows/apps/example/gif/giflist/GifAdapterTest.kt
package burrows.apps.example.gif.giflist

import android.support.test.InstrumentationRegistry
import android.support.test.annotation.UiThreadTest
import android.view.ViewGroup
import android.widget.LinearLayout
import burrows.apps.example.gif.data.ImageService
import burrows.apps.example.gif.giflist.GifAdapter.OnItemClickListener
import com.google.common.truth.Truth.assertThat
import org.junit.Before
import org.junit.Test

class GifAdapterTest {
    private val targetContext = InstrumentationRegistry.getTargetContext()
    private val gifImageInfo = GifImageInfo(url = "http://some.url")
    private val gifImageInfo2 = GifImageInfo(url = "http://some.url2")
    private val gifImageInfo3 = GifImageInfo(url = "http://some.url3")
    private val testOnItemClickListener = object : OnItemClickListener {
        override fun onClick(imageInfoModel: GifImageInfo) {
        }
    }
    private lateinit var imageService: ImageService
    private lateinit var viewHolder: GifAdapter.ViewHolder
    private lateinit var sut: GifAdapter

    @Before @UiThreadTest fun setUp() {
        imageService = ImageService(targetContext)
        sut = GifAdapter(testOnItemClickListener, imageService).apply {
            add(gifImageInfo)
            add(gifImageInfo2)
        }
        viewHolder = sut.onCreateViewHolder(LinearLayout(targetContext), 0)
    }

    @Test @UiThreadTest fun testOnCreateViewHolder() {
        val parent = object : ViewGroup(targetContext) {
            override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {}
        }

        assertThat(sut.onCreateViewHolder(parent, 0))
            .isInstanceOf(GifAdapter.ViewHolder::class.java)
    }

    @Test @UiThreadTest fun testOnBindViewHolderOnAdapterItemClick() {
        sut.clear()
        sut.add(gifImageInfo)
        sut.add(gifImageInfo2)
        sut.add(GifImageInfo())

        sut.onBindViewHolder(viewHolder, 0)

        assertThat(viewHolder.itemView.performClick()).isTrue()
    }

    @Test fun testGetItem() {
        sut.clear()

        val imageInfo = GifImageInfo()
        sut.add(imageInfo)

        assertThat(sut.getItem(0)).isEqualTo(imageInfo)
    }

    @Test @UiThreadTest fun onViewRecycled() {
        sut.add(GifImageInfo())

        sut.onBindViewHolder(viewHolder, 0)
        sut.onViewRecycled(viewHolder)
    }

    @Test fun testGetItemCountShouldReturnCorrectValues() {
        assertThat(sut.itemCount).isEqualTo(2)
    }

    @Test fun testGetListCountShouldReturnCorrectValues() {
        assertThat(sut.getItem(0)).isEqualTo(gifImageInfo)
        assertThat(sut.getItem(1)).isEqualTo(gifImageInfo2)
    }

    @Test fun testGetItemShouldReturnCorrectValues() {
        assertThat(sut.getItem(1)).isEqualTo(gifImageInfo2)
    }

    @Test fun testClearShouldClearAdapter() {
        sut.clear()

        assertThat(sut.itemCount).isEqualTo(0)
    }

    @Test fun testAddObjectShouldReturnCorrectValues() {
        sut.add(gifImageInfo3)

        assertThat(sut.getItem(0)).isEqualTo(gifImageInfo)
        assertThat(sut.getItem(1)).isEqualTo(gifImageInfo2)
        assertThat(sut.getItem(2)).isEqualTo(gifImageInfo3)
    }

    @Test fun testAddCollectionShouldReturnCorrectValues() {
        val imageInfos = listOf(gifImageInfo3)

        sut.addAll(imageInfos)

        assertThat(sut.getItem(0)).isEqualTo(gifImageInfo)
        assertThat(sut.getItem(1)).isEqualTo(gifImageInfo2)
        assertThat(sut.getItem(2)).isEqualTo(gifImageInfo3)
    }

    @Test fun testAddLocationObjectShouldReturnCorrectValues() {
        sut.add(0, gifImageInfo3)

        assertThat(sut.getItem(0)).isEqualTo(gifImageInfo3)
        assertThat(sut.getItem(1)).isEqualTo(gifImageInfo)
        assertThat(sut.getItem(2)).isEqualTo(gifImageInfo2)
    }
}
